### PR TITLE
X509: use V3 when extensions present; certs generated by apps bear X.509 V3 by default

### PR DESCRIPTION
**Pull request by DDvO**
When preparing #19076 I found that by default (when no X.509 v3 extensions are used),
certs generated by the `x509`, `ca`, and `req` apps still have X.509 version 1.
This is anachronistic and causes various issues, such as missing key identifiers for chain building
and no clear distinction between CA certs and EE certs.

This PR makes sure V3 is used, 
with the exception of a new `-x509v1` option of the `req` app, which implies the `-x509` option while ~forcing~ using V1 unless extensions are given (which requires V3).

On this occasion, also
* `apps/req.c`: properly report parse errors by `duplicated()`; simplify the function
* ~`CHANGES.md:` 'added' -> 'add', 'extended' -> 'extend' for OpenSSL 3.1 entries~
* ~`test/trace_api_test.c`: fix gcc error on `-Werror=strict-prototypes`~

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by DDvO** [comment]
The failing `external-tests` CI run with `error 79 at 1 depth lookup: invalid CA certificate`
are most likely due to missing `basicConstraints = CA:true`, which I also fixed in `90-test_store.t`.
@beldmit, please verify and fix.

**Comment by DDvO** [comment]
And how about the API level?

I suggest adapting `X509_sign{_ctx}()` to set version 3 if extensions are present,
which prevents `X509_V_ERR_EXTENSIONS_REQUIRE_VERSION_3` on verification.

**Comment by beldmit** [comment]
@DDvO could you please try the https://github.com/gost-engine/engine/commit/a6b90523e4ea6010b1109b0bae7e2a73b5b025c5 ?

**Comment by DDvO** [comment]
> @DDvO could you please try the [gost-engine/engine@a6b9052](https://github.com/gost-engine/engine/commit/a6b90523e4ea6010b1109b0bae7e2a73b5b025c5) ?

Thank you for your immediate handling - with this updated submodule reference it works fine :)

**Review by t8m — CHANGES_REQUESTED** [review]
I would drop the first commit. IMO the past tense is correct.

File: test/trace_api_test.c [hunk 1]
```diff
@@ -67,15 +67,15 @@ static int test_trace_categories(void)
 }
 
 #ifndef OPENSSL_NO_TRACE
-static void put_trace_output()
+static void put_trace_output(void)
```
Comment (t8m) [inline test/trace_api_test.c:70]: Please do not put unrelated changes into a PR.

File: apps/lib/apps.c [hunk 2]
```diff
@@ -2288,16 +2288,14 @@ int cert_matches_key(const X509 *cert, const EVP_PKEY *pkey)
...
                  STACK_OF(OPENSSL_STRING) *sigopts, X509V3_CTX *ext_ctx)
 {
-    const STACK_OF(X509_EXTENSION) *exts = X509_get0_extensions(cert);
     EVP_MD_CTX *mctx = EVP_MD_CTX_new();
     int self_sign;
     int rv = 0;
 
-    if (sk_X509_EXTENSION_num(exts /* may be NULL */) > 0) {
-        /* Prevent X509_V_ERR_EXTENSIONS_REQUIRE_VERSION_3 */
+    if (!force_v1) {
```
Comment (t8m) [inline apps/lib/apps.c:2298]: Should you also check that sk_X509_EXTENSION_num(exts) == 0 if force_v1 is set?

File: test/trace_api_test.c (hunk 1 above)
Comment (DDvO) [reply test/trace_api_test.c:70]: I was compelled to include this because otherwise my local builds fail with:
```
test/trace_api_test.c:70:13: error: function declaration isn't a prototype [-Werror=strict-prototypes]
   70 | static void put_trace_output()
      |             ^~~~~~~~~~~~~~~~
test/trace_api_test.c:78:12: error: function declaration isn't a prototype [-Werror=strict-prototypes]
   78 | static int test_trace_channel()
      |            ^~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```
since I generally configure with `--strict-warnings -std=gnu90`.
I fear that some CI part(s) will also fail without those two fixes.

File: apps/lib/apps.c (hunk 2 above)
Comment (DDvO) [reply apps/lib/apps.c:2298]: I thought about this, too.
Allowing for `sk_X509_EXTENSION_num(exts) > 0` to override `force_v1` will require tuning down the name and description of `force_v1` and the `-x509v1` option.
And I believe this should be more generally done at libcrypto level., as I suggested above:
Whenever in the end just before signing a cert there are some extensions, promote the cert to V3.

**Comment by DDvO** [comment]
> I would drop the first commit. IMO the past tense is correct.

Will do with the next push.

File: apps/lib/apps.c (hunk 2 above)
Comment (DDvO) [reply apps/lib/apps.c:2298]: Shall I weaken `force_v1` like this
and do the suggested change in `X509_sign{_ctx}()` (to set version 3 if extensions are present)?
This would make sure that certs bear version 3 whenever needed.

File: test/trace_api_test.c (hunk 1 above)
Comment (DDvO) [reply test/trace_api_test.c]: Rebased on the meanwhile merged PR #19277, so that commit is no more part of this PR.

File: apps/lib/apps.c (hunk 2 above)
Comment (DDvO) [reply apps/lib/apps.c:2298]: Done.

**Comment by DDvO** [comment]
@t8m, ok now?

**Review by t8m — CHANGES_REQUESTED** [review]
Just a few nits to fix

File: CHANGES.md [hunk 3]
```diff
@@ -155,6 +155,13 @@ OpenSSL 3.1
 
    *Tomáš Mráz*
 
+ * The `x509`, `ca`, and `req` apps now produce X.509 v3 certs.
```
[the remark this answers is not in the document]
Comment (DDvO) [reply CHANGES.md]: Changed

File: CHANGES.md [hunk 4]
```diff
@@ -155,6 +155,13 @@ OpenSSL 3.1
...
+   The `-x509v1` option of the `req` app requests generation of X.509 v1 certs.
+   `X509_sign()` and `X509_sign_ctx()` make sure that the certficiate has
+   X.509 version 3 if the  certificate information includes X.509 extensions.
```
[the remark this answers is not in the document]
Comment (DDvO) [reply CHANGES.md]: Removed.

File: CHANGES.md [hunk 5]
```diff
@@ -155,6 +155,13 @@ OpenSSL 3.1
 
    *Tomáš Mráz*
 
+ * The `x509`, `ca`, and `req` apps now produce X.509 v3 certs.
+   The `-x509v1` option of the `req` app requests generation of X.509 v1 certs.
+   `X509_sign()` and `X509_sign_ctx()` make sure that the certficiate has
```
[the remark this answers is not in the document]
Comment (DDvO) [reply CHANGES.md]: Oops, my favorite typo - fixed.

**Comment by DDvO** [comment]
> Just a few nits to fix

Thanks @t8m for your further comments, which I had overlooked so far, but now they are handled.

**Review by t8m — APPROVED** [review]
CI failures are unrelated.

**Comment by DDvO** [comment]
Rebased to fix merge conflict in `x_all.c`

Ping for 2nd review.

**Comment by DDvO** [comment]
Rebased to fix trivial merge conflict in `CHANGES.md`.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Adding some more tests would be nice.

**Comment by DDvO** [comment]
> Adding some more tests would be nice.

Done.

**Comment by DDvO** [comment]
@t8m wanna reconfirm, now since more tests have been added?
I also added note to the HISTORY section of the three app docs.

**Review by t8m — APPROVED** [review]
CI failure is not related.

**Comment by t8m** [comment]
ping for second review

**Review by hlandau — APPROVED** [review]

**Comment by DDvO** [comment]
Merged  - thanks @t8m and @hlandau for the reviews.
